### ISSUE TITLE
[xml] New language to detect issues in Salesforce Flow

### ISSUE THREAD

**Issue by aidan-harding** [pmd/pmd#3413]
**Is your feature request related to a problem? Please describe.**

Salesforce Flow is a visual coding tool. It offers normal programming constructs such as loops and branching and it is stored in an XML format. Many of the rules applicable to Salesforce Apex also apply to Flow e.g. no database operations within loops, no use of hard-coded database record ids etc. 

It would be of great benefit to Flow developers to have PMD as a guide to help them in writing efficient, safe, and maintainable Flows. It also nicely complements the existing Salesforce Apex language.

**Describe the solution you'd like**

A new language option for Flow (or perhaps a new language for Salesforce in general that would allow checking of all languages within Salesforce from a single option?) which parses the XML runs appropriate rules.

I have already started a fork to work on this myself where I run PMD with these flags:

```
-d ../flow-pmd/force-app/main/default/flows/ -R src/main/resources/category/salesforceFlow/bestpractices.xml -language salesforceFlow
```

The rules are in Java because the XML structure used by Flow does not reflect the programmatic structure of the Flow e.g. if the Flow contains a loop, the XML does not have the body of the loop inside the XML loop node. Instead, links between nodes are referenced via names. So, the Java implementation uses the XML to build a graph corresponding to the Flow as it would be executed, then we can write rules on that graph.

**Describe alternatives you've considered**

As above, the language could be a single one for Salesforce Flow, or an overarching Salesforce language allowing expansion to check other Salesforce items without hugely expanding the language list. I don't know the internals of PMD well enough to know the best course there.

I also considered transpiling the Flow XML format into another XML format which more accurately reflects the graph of the Flow to be executed. This would allow us to write rules on the transpiled XML in XPath. That would be neat, but possibly confusing without being able to see the transpiled XML. 

**Additional context**

An overview of Flow: https://www.salesforceben.com/introduction-salesforce-flow/

My fork is here (although I haven't pushed my local changes to it yet) https://github.com/aidan-harding/pmd

**Comment by aidan-harding** [comment]
Note that this is related to #2133, but not the same. Back then, Salesforce had a different language, Process Builder, for this visual programming. That language used files that did not end in .XML, so it tripped up over #1540. The proposal above is possible without fixing #1540 because the files do end in .XML.

**Comment by rsoesemann** [comment]
Thanks, @aidan-harding for tackling this finally. Many Salesforce Developers are waiting for using PMD to check more metadata types. Especially one that is getting more important every day like Salesforce Flow.

**Comment by aidan-harding** [comment]
The more I look at this, the more I think it will be best to translate the Flow into its own AST. If we do that, then it will show up in the designer tool, enabling XPath rules easily. That's actually close to what I was doing anyway, I just need to plug my graph nodes into the PMD API. 

I fear that I will then have to create my own language and sort out separating language choice from the file extension, but that doesn't actually look very hard.

**Comment by rsoesemann** [comment]
@jbartolotta-sfdc @rmohan20 I just wanted to draw your attention to @aidan-harding contributing this. It opens doors to have PMD check arbitrary XML Metadata (MDAPI and DX). And people can use custom XPath rules in their local rulesets without becoming Java rule contributor. A big blocker for PMD adoption (only Apex) and contribution (on OSS via Java) will be resolved with that.

**Comment by jbartolotta-sfdc** [comment]
@aidan-harding this will be a great addition to PMD. 

@rmohan20 and I work for Salesforce(not on the Flow team), please reach out if you have questions and we can work to get you answers.

**Comment by oowekyala** [comment]
> The more I look at this, the more I think it will be best to translate the Flow into its own AST. If we do that, then it will show up in the designer tool, enabling XPath rules easily. That's actually close to what I was doing anyway, I just need to plug my graph nodes into the PMD API.

> I fear that I will then have to create my own language and sort out separating language choice from the file extension, but that doesn't actually look very hard.

If you want to use a separate AST, then yes, you'll have to create a new language. But this can let you implement new xpath functions with language-specific semantics, which is nice.

Until then, with #3417 you'll be able to write `--force-language xml` and use the default XML xpath support

**Comment by SCWells72** [comment]
Moving this from #3846 since it's specific to this issue. Regarding the provided Flow XML and wanting to detect DML in loops, you can query such nodes as follows:

```xml
//Flow/loops[nextValueConnector/targetReference = //Flow/recordUpdates/name]
```

That basically says "find a loop where the target is a record update", and the `loops` node is selected. Is that not what you want?

**Comment by SCWells72** [comment]
As a slight update, I added the following rule to my PMD ruleset:

```xml
    <rule name="FlowNoDmlInLoops" language="xml" message="DML should not be performed in loops." class="net.sourceforge.pmd.lang.rule.XPathRule">
        <priority>3</priority>
        <properties>
            <property name="version" value="2.0"></property>
            <property name="xpath">
                <value><![CDATA[
                //Flow/loops[nextValueConnector/targetReference = //Flow/recordUpdates/name]
                ]]> </value>
            </property>
        </properties>
    </rule>
```

and, in a project which contains the Flow file you provided on #3846, I get the following violation:

<img width="518" alt="image" src="https://user-images.githubusercontent.com/7671043/158862079-c559bc5e-6623-4bae-ba84-0e97210638ce.png">

That can be generalized further by ORing the selection criteria for check for other DML nodes such as `recordCreates` and `recordDeletes`.

**Comment by RubenHalman** [comment]
But this would only look at the first next node after the loop starts, am I right? The delay statement can the 3th node in the loop and it wouldn't be flagged with this rule.

**Comment by SCWells72** [comment]
I'll be away from my machine for a few hours, but I'm pretty sure there's a way to specify repetition in paths so that it can handle arbitrary levels of separation. I'll check into it in a bit and see if I can provide a concrete rule that does so.

**Comment by SCWells72** [comment]
What I provided before was incorrect, but I believe something very much like the following should work (but doesn't):

```xpath
/Flow/loops//targetReference/text[@Image = /Flow/recordUpdates/name/text[@Image]]
```

My best understanding is that effectively says that it should find any element under `/Flow/loops` named `targetReference` with arbitrary intermediate paths (the `//`) with a text value that matches any other element in the same document under `/Flow/recordUpdates` with a `name` element with the same value.

To be clear, when I enter either of these paths in isolation into the PMD rule designer, the expected nodes' text values are selected:

```xpath
/Flow/loops//targetReference/text[@Image]
/Flow/recordUpdates/name/text[@Image]
```

but when I enter the full expression, I don't see the expected nodes selected.

To be clear, this is getting to the far reaches of my XPath-Fu, but the problem we're trying to solve here is virtually identical to this:

https://stackoverflow.com/questions/27151688/xpath-select-node-based-on-other-nodes-value

except for wanting arbitrary intermediate paths which should be enabled by the `//`.

I'm pretty sure this is very close to being the answer, and then of course you just have to adjust the `recordUpdates` to select any of the DML-type elements for that part of the predicate.

I'm happy to work on this a bit more next week, but I won't get much chance to do so before then most likely as I'll be out of pocket Friday-Sunday. Certainly if anyone else has the time and/or know-how to complete this thought, even better...

**Comment by RubenHalman** [comment]
Thank you, its most interesting Scott! A challenge is that the dml statement can be outside of the loop element, I'll be looking forward to if and how we can do it!

**Comment by aidan-harding** [comment]
Here's the comment I posted in the other issue:

@SCWells72 Here you go:

![image](https://user-images.githubusercontent.com/35935718/158853331-7798e480-6be1-463d-b70b-20025859dd44.png)

That Flow is represented by:

```
<?xml version="1.0" encoding="UTF-8"?>
<Flow xmlns="http://soap.sforce.com/2006/04/metadata">
    <apiVersion>54.0</apiVersion>
    <assignments>
        <name>Set_Contact_Description</name>
        <label>Set Contact Description</label>
        <locationX>264</locationX>
        <locationY>695</locationY>
        <assignmentItems>
            <assignToReference>Loop_Contacts.Description</assignToReference>
            <operator>Assign</operator>
            <value>
                <stringValue>Updated</stringValue>
            </value>
        </assignmentItems>
        <connector>
            <targetReference>Loop_Contacts</targetReference>
        </connector>
    </assignments>
    <interviewLabel>Account Update Contacts In Loop {!$Flow.CurrentDateTime}</interviewLabel>
    <label>Account Update Contacts In Loop</label>
    <loops>
        <name>Loop_Contacts</name>
        <label>Loop Contacts</label>
        <locationX>176</locationX>
        <locationY>455</locationY>
        <collectionReference>Get_Contacts</collectionReference>
        <iterationOrder>Asc</iterationOrder>
        <nextValueConnector>
            <targetReference>Update_Contact</targetReference>
        </nextValueConnector>
    </loops>
    <processMetadataValues>
        <name>BuilderType</name>
        <value>
            <stringValue>LightningFlowBuilder</stringValue>
        </value>
    </processMetadataValues>
    <processMetadataValues>
        <name>CanvasMode</name>
        <value>
            <stringValue>AUTO_LAYOUT_CANVAS</stringValue>
        </value>
    </processMetadataValues>
    <processMetadataValues>
        <name>OriginBuilderType</name>
        <value>
            <stringValue>LightningFlowBuilder</stringValue>
        </value>
    </processMetadataValues>
    <processType>AutoLaunchedFlow</processType>
    <recordLookups>
        <name>Get_Contacts</name>
        <label>Get Contacts</label>
        <locationX>176</locationX>
        <locationY>335</locationY>
        <assignNullValuesIfNoRecordsFound>false</assignNullValuesIfNoRecordsFound>
        <connector>
            <targetReference>Loop_Contacts</targetReference>
        </connector>
        <filterLogic>and</filterLogic>
        <filters>
            <field>AccountId</field>
            <operator>EqualTo</operator>
            <value>
                <elementReference>$Record.Id</elementReference>
            </value>
        </filters>
        <getFirstRecordOnly>false</getFirstRecordOnly>
        <object>Contact</object>
        <storeOutputAutomatically>true</storeOutputAutomatically>
    </recordLookups>
    <recordUpdates>
        <name>Update_Contact</name>
        <label>Update Contact</label>
        <locationX>264</locationX>
        <locationY>575</locationY>
        <connector>
            <targetReference>Set_Contact_Description</targetReference>
        </connector>
        <inputReference>Loop_Contacts</inputReference>
    </recordUpdates>
    <start>
        <locationX>50</locationX>
        <locationY>0</locationY>
        <connector>
            <targetReference>Get_Contacts</targetReference>
        </connector>
        <object>Account</object>
        <recordTriggerType>Update</recordTriggerType>
        <triggerType>RecordAfterSave</triggerType>
    </start>
    <status>Draft</status>
</Flow>
```

So you can see that the update being inside the loop is represented by the indirect reference `<targetReference>Update_Contact</targetReference>`. Which is not ideal.

**Comment by adangel** [comment]
> `/Flow/loops//targetReference/text[@Image=/Flow/(recordCreates|recordDeletes|recordUpdates)/name/text[@Image]]`

With the square-brackets at the end, xpath tries to compare the attribute "Image" with any element "text", that has an attribute "Image". So, it tries to compare an attribute with an element, which evaluates always to false.

I think, this should be

```
/Flow/loops//targetReference/text[@Image=/Flow/(recordCreates|recordDeletes|recordUpdates)/name/text/@Image]
```

so that `@Image`-attributes are compared.

**Comment by oowekyala** [comment]
> Notice the `//Flow` in the filter. Is that perhaps needed because otherwise the XPath expression inside the square brackets is being evaluated relative to the outer path?

No, this selects `Flow` nodes in the tree, while `/Flow` selects only the root node if it is named Flow. Since your document root is a `Flow`, `/Flow` should match it, so there must be something wrong with your AST,  (or you're using XPath 1, refs #1938).


You could be creating custom XPath functions to simplify this. Eg you could create a `resolveTarget` function that takes a `loops` as a parameter and returns the next node.
```xml
/Flow/loops[
  let $target := pmd-flow:resolveTarget(.)
  (: or use $target in whatever way. :)
  return $target/@Image ]
```

The complexity of resolving indirect references and such would be hidden inside the function.

**Comment by SCWells72** [comment]
@oowekyala this behavior is all from the PMD rule designer. Here it is with a second path that starts with `//Flow` where the node is selected:

![image](https://user-images.githubusercontent.com/7671043/159058374-4f49bbca-5abc-47f9-9d0b-d27a0d22d513.png)

and here it is with just `/Flow` where no node is selected:

![image](https://user-images.githubusercontent.com/7671043/159058436-b8ba9d7b-aeb7-4a0a-83c0-3c4d51e19224.png)

Is there something else I need to do in the designer? Or has this unveiled a bug in PMD's XPath functionality?

@aidan-harding and @RubenHalman, it's probably worth looking into the XPath function approach suggested above as an alternative to a full language/AST to resolve references which can't be directly modeled in XPath itself. There's obviously a complexity trade-off. I think that if very good examples are provided of the use of these functions in the default Salesforce ruleset, specifically for this use case of identifying DML in loops, the lower overall investment in a few XPath functions vs. a completely new PMD language could well tip the balance.

**Comment by RubenHalman** [comment]
That makes sense. This is why I solved it with java script instead of xpath for the lightning flow scanner. Let me know how I can help or maybe have a look at how I did it with Js

https://marketplace.visualstudio.com/items?itemName=ForceConfigControl.lightningflowscanner

**Comment by oowekyala** [comment]
> Is there something else I need to do in the designer? Or has this unveiled a bug in PMD's XPath functionality?

Thanks for trying this out. I believe this is a bug in our AST wrapper for the XML DOM, there shouldn't be an explicit node named "document" as the root.

I think it would be nice if XPath rules for the XML language behaved exactly as regular XPath does. This is related to this bug (#915), where the standard XPath function `text()` doesn't work, because our wrapper instead creates a node `text` with an `@Image`. I think this is completely wrong and XML xpath rules would be more powerful and easier to write if we used Saxon's own xpath engine on the DOM instead of our AST wrapper. I think, to fix this, we need a new rule class `XmlXPathRule` that invokes Saxon directly (or rather, we should rehabilitate the current one, currently deprecated).


I'll be happy to provide guidance about implementing custom xpath functions to anyone interested

**Comment by adangel** [comment]
Here's a xpath rexpression, that can find indirect calls up to 3 indirections. It's of course not nice and hiding this in a function would be cleaner, but maybe 3 indirections are enough for the beginning?

```
(
    (: direct reference :)
    //Flow/loops[nextValueConnector/targetReference/text/@Image=//Flow/(recordCreates|recordDeletes|recordUpdates)/name/text/@Image]
    |
    (: indirect reference :)
    (for $node in //Flow/loops,
         $target1 in //Flow/*[name/text/@Image=$node//targetReference/text/@Image],
         $target2 in //Flow/(recordCreates|recordDeletes|recordUpdates)[name/text/@Image=$target1//targetReference/text/@Image]
            return $node
    )
    |
    (: indirect reference, 2nd level :)
    (for $node in //Flow/loops,
         $target1 in //Flow/*[name/text/@Image=$node//targetReference/text/@Image],
         $target2 in //Flow/*[name/text/@Image=$target1//targetReference/text/@Image],
         $target3 in //Flow/(recordCreates|recordDeletes|recordUpdates)[name/text/@Image=$target2//targetReference/text/@Image]
            return $node
    )
    |
    (: indirect reference, 3rd level :)
    (for $node in //Flow/loops,
         $target1 in //Flow/*[name/text/@Image=$node//targetReference/text/@Image],
         $target2 in //Flow/*[name/text/@Image=$target1//targetReference/text/@Image],
         $target3 in //Flow/*[name/text/@Image=$target2//targetReference/text/@Image],
         $target4 in //Flow/(recordCreates|recordDeletes|recordUpdates)[name/text/@Image=$target3//targetReference/text/@Image]
              return $node
    )
)/name

```

**Comment by tomekduda** [comment]
I'd like to propose a modification but my XPath skills are rusty.

This test case
![image](https://user-images.githubusercontent.com/7957690/193430818-627b3d7d-f9ef-457a-8f18-a02a936e8e36.png)

Is represented by

<details>
  <summary>Click to expand</summary>
  
  ```xml
  <?xml version="1.0" encoding="UTF-8"?>
<Flow xmlns="http://soap.sforce.com/2006/04/metadata">
    <actionCalls>
        <description>You can&apos;t tell here but it runs a query and we call it in a loop...</description>
        <name>call_Apex_code</name>
        <label>call Apex code</label>
        <locationX>794</locationX>
        <locationY>695</locationY>
        <actionName>AccountQueryAction</actionName>
        <actionType>apex</actionType>
        <connector>
            <targetReference>suspiciousLoop1</targetReference>
        </connector>
        <flowTransactionModel>CurrentTransaction</flowTransactionModel>
        <storeOutputAutomatically>true</storeOutputAutomatically>
    </actionCalls>
    <apiVersion>55.0</apiVersion>
    <assignments>
        <name>cascadeCity1</name>
        <label>cascadeCity1</label>
        <locationX>138</locationX>
        <locationY>695</locationY>
        <assignmentItems>
            <assignToReference>badLoop1.MailingCity</assignToReference>
            <operator>Assign</operator>
            <value>
                <elementReference>$Record.BillingCity</elementReference>
            </value>
        </assignmentItems>
        <connector>
            <targetReference>badUpdate1</targetReference>
        </connector>
    </assignments>
    <assignments>
        <name>cascadeCity2</name>
        <label>cascadeCity2</label>
        <locationX>442</locationX>
        <locationY>695</locationY>
        <assignmentItems>
            <assignToReference>goodLoop1.MailingCity</assignToReference>
            <operator>Assign</operator>
            <value>
                <elementReference>$Record.BillingCity</elementReference>
            </value>
        </assignmentItems>
        <connector>
            <targetReference>goodLoop1</targetReference>
        </connector>
    </assignments>
    <decisions>
        <description>just to see different paths side by side. If it&apos;d be 1 long flow the false positives could kick in</description>
        <name>whichCity</name>
        <label>whichCity</label>
        <locationX>378</locationX>
        <locationY>455</locationY>
        <defaultConnector>
            <targetReference>suspiciousLoop1</targetReference>
        </defaultConnector>
        <defaultConnectorLabel>Default Outcome</defaultConnectorLabel>
        <rules>
            <name>which_city</name>
            <conditionLogic>and</conditionLogic>
            <conditions>
                <leftValueReference>$Record.BillingCity</leftValueReference>
                <operator>EqualTo</operator>
                <rightValue>
                    <stringValue>London</stringValue>
                </rightValue>
            </conditions>
            <connector>
                <targetReference>badLoop1</targetReference>
            </connector>
            <label>is London</label>
        </rules>
        <rules>
            <name>isNewYork</name>
            <conditionLogic>and</conditionLogic>
            <conditions>
                <leftValueReference>$Record.BillingCity</leftValueReference>
                <operator>EqualTo</operator>
            </conditions>
            <connector>
                <targetReference>goodLoop1</targetReference>
            </connector>
            <label>isNewYork</label>
        </rules>
    </decisions>
    <description>For detecting database access in a loop</description>
    <environments>Default</environments>
    <interviewLabel>Sample PMD flow {!$Flow.CurrentDateTime}</interviewLabel>
    <label>Sample PMD flow</label>
    <loops>
        <description>Contains update in a loop, should be caught</description>
        <name>badLoop1</name>
        <label>badLoop1</label>
        <locationX>50</locationX>
        <locationY>575</locationY>
        <collectionReference>getContacts</collectionReference>
        <iterationOrder>Asc</iterationOrder>
        <nextValueConnector>
            <targetReference>cascadeCity1</targetReference>
        </nextValueConnector>
        <noMoreValuesConnector>
            <targetReference>X1_more_update</targetReference>
        </noMoreValuesConnector>
    </loops>
    <loops>
        <description>This does update outside of the loop, should be good</description>
        <name>goodLoop1</name>
        <label>goodLoop1</label>
        <locationX>354</locationX>
        <locationY>575</locationY>
        <collectionReference>getContacts</collectionReference>
        <iterationOrder>Asc</iterationOrder>
        <nextValueConnector>
            <targetReference>cascadeCity2</targetReference>
        </nextValueConnector>
        <noMoreValuesConnector>
            <targetReference>goodUpdate1</targetReference>
        </noMoreValuesConnector>
    </loops>
    <loops>
        <description>Calling Apex in a loop is smelly. Why do you need to bail out and write code. There&apos;s a chance there are inserts, updates, maybe callouts (@future but still). It&apos;s cleaner to call apex once, outside of the loop, passing a collection as input</description>
        <name>suspiciousLoop1</name>
        <label>suspiciousLoop1</label>
        <locationX>706</locationX>
        <locationY>575</locationY>
        <collectionReference>getContacts</collectionReference>
        <iterationOrder>Asc</iterationOrder>
        <nextValueConnector>
            <targetReference>call_Apex_code</targetReference>
        </nextValueConnector>
        <noMoreValuesConnector>
            <targetReference>X1_more_update</targetReference>
        </noMoreValuesConnector>
    </loops>
    <processMetadataValues>
        <name>BuilderType</name>
        <value>
            <stringValue>LightningFlowBuilder</stringValue>
        </value>
    </processMetadataValues>
    <processMetadataValues>
        <name>CanvasMode</name>
        <value>
            <stringValue>AUTO_LAYOUT_CANVAS</stringValue>
        </value>
    </processMetadataValues>
    <processMetadataValues>
        <name>OriginBuilderType</name>
        <value>
            <stringValue>LightningFlowBuilder</stringValue>
        </value>
    </processMetadataValues>
    <processType>AutoLaunchedFlow</processType>
    <recordLookups>
        <description>This is ok, outside of loop</description>
        <name>getContacts</name>
        <label>getContacts</label>
        <locationX>378</locationX>
        <locationY>335</locationY>
        <assignNullValuesIfNoRecordsFound>false</assignNullValuesIfNoRecordsFound>
        <connector>
            <targetReference>whichCity</targetReference>
        </connector>
        <filterLogic>and</filterLogic>
        <filters>
            <field>AccountId</field>
            <operator>EqualTo</operator>
            <value>
                <elementReference>$Record.Id</elementReference>
            </value>
        </filters>
        <filters>
            <field>MailingCity</field>
            <operator>NotEqualTo</operator>
            <value>
                <elementReference>$Record.BillingCity</elementReference>
            </value>
        </filters>
        <getFirstRecordOnly>false</getFirstRecordOnly>
        <object>Contact</object>
        <storeOutputAutomatically>true</storeOutputAutomatically>
    </recordLookups>
    <recordUpdates>
        <name>badUpdate1</name>
        <label>badUpdate1</label>
        <locationX>138</locationX>
        <locationY>815</locationY>
        <connector>
            <targetReference>badLoop1</targetReference>
        </connector>
        <inputReference>badLoop1</inputReference>
    </recordUpdates>
    <recordUpdates>
        <name>goodUpdate1</name>
        <label>goodUpdate1</label>
        <locationX>354</locationX>
        <locationY>911</locationY>
        <connector>
            <targetReference>X1_more_update</targetReference>
        </connector>
        <inputReference>goodLoop1</inputReference>
    </recordUpdates>
    <recordUpdates>
        <name>X1_more_update</name>
        <label>1 more update</label>
        <locationX>378</locationX>
        <locationY>1127</locationY>
        <inputAssignments>
            <field>Description</field>
            <value>
                <stringValue>x</stringValue>
            </value>
        </inputAssignments>
        <inputReference>$Record</inputReference>
    </recordUpdates>
    <start>
        <locationX>252</locationX>
        <locationY>0</locationY>
        <connector>
            <targetReference>getContacts</targetReference>
        </connector>
        <filterLogic>and</filterLogic>
        <filters>
            <field>BillingCity</field>
            <operator>IsChanged</operator>
            <value>
                <booleanValue>true</booleanValue>
            </value>
        </filters>
        <object>Account</object>
        <recordTriggerType>Update</recordTriggerType>
        <triggerType>RecordAfterSave</triggerType>
    </start>
    <status>Draft</status>
</Flow>
  ```
</details>

@adangel's rule flags all 3 loops but it shouldn't

1. should be flagged
2. false positive, the update is outside of the loop
3. should be flagged but for different reason. the update is safe, outside of the loop, the dangerous element is new

I think it's because loop tags have 2x `targetReference` and XPath will evaluate both `nextValueConnector` and `noMoreValuesConnector`?

Also I'd like to suggest detecting calls to Apex classes or other flows in a loop. Lower priority but flagging potentially suspicious areas. It can be innocent (things like do some hardcore string manipulation and return for example would be painful to express in the flow) - but most of the time I see "invocable actions" used it involves database.

I managed to get this far.

<details>
  <summary>Click to expand</summary>
  
  ```xml
<ruleset
    xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    name="Salesforce Flow"
    xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 http://pmd.sourceforge.net/ruleset_2_0_0.xsd">
    <description>Salesforce Flow PMD ruleset, focused on performance issues</description>
    <exclude-pattern>.*/force-app/main/default/.*</exclude-pattern>
    <include-pattern>.*/force-app/main/default/flows/.*</include-pattern>

    <rule
        name="FlowWithSoqlDmlInLoop"
        language="xml"
        message="Avoid SOQL, DML in a loop in flows"
        class="net.sourceforge.pmd.lang.rule.XPathRule">
        <description/>
        <priority>1</priority>
        <properties>
            <property name="version" value="2.0"/>
            <property name="xpath">
                <value>
                <![CDATA[
                (
                    (: direct reference :)
                    //Flow/loops[nextValueConnector/targetReference/text/@Image=//Flow/(recordLookups|recordCreates|recordDeletes|recordUpdates)/name/text/@Image]
                    |
                    (: indirect reference :)
                    (for $node in //Flow/loops,
                        $target1 in //Flow/*[name/text/@Image=$node//nextValueConnector/targetReference/text/@Image]
                            return $node[//Flow/(recordLookups|recordCreates|recordDeletes|recordUpdates)[name/text/@Image=$target1//targetReference/text/@Image]]
                    )
                    |
                    (: indirect reference, 2nd level :)
                    (for $node in //Flow/loops,
                        $target1 in //Flow/*[name/text/@Image=$node//nextValueConnector/targetReference/text/@Image],
                        $target2 in //Flow/*[name/text/@Image=$target1//nextValueConnector/targetReference/text/@Image]
                            return $node[//Flow/(recordLookups|recordCreates|recordDeletes|recordUpdates)[name/text/@Image=$target2//targetReference/text/@Image]]
                    )
                    |
                    (: indirect reference, 3rd level :)
                    (for $node in //Flow/loops,
                        $target1 in //Flow/*[name/text/@Image=$node//nextValueConnector/targetReference/text/@Image],
                        $target2 in //Flow/*[name/text/@Image=$target1//nextValueConnector/targetReference/text/@Image],
                        $target3 in //Flow/*[name/text/@Image=$target2//nextValueConnector/targetReference/text/@Image]
                            return $node[//Flow/(recordLookups|recordCreates|recordDeletes|recordUpdates)[name/text/@Image=$target3//targetReference/text/@Image]]
                    )
                )/name
                ]]>
                </value>
            </property>
        </properties>
    </rule>
    <rule
        name="FlowWithApexInLoop"
        language="xml"
        message="Avoid Invocable Actions in a loop in flows"
        class="net.sourceforge.pmd.lang.rule.XPathRule">
        <description/>
        <priority>2</priority>
        <properties>
            <property name="version" value="2.0"/>
            <property name="xpath">
                <value>
                <![CDATA[
                (
                    (: direct reference :)
                    //Flow/loops[nextValueConnector/targetReference/text/@Image=//Flow/actionCalls/name/text/@Image]
                    |
                    (: indirect reference :)
                    (for $node in //Flow/loops,
                        $target1 in //Flow/*[name/text/@Image=$node//nextValueConnector/targetReference/text/@Image]
                            return $node[//Flow/actionCalls[name/text/@Image=$target1//targetReference/text/@Image]]
                    )
                    |
                    (: indirect reference, 2nd level :)
                    (for $node in //Flow/loops,
                        $target1 in //Flow/*[name/text/@Image=$node//nextValueConnector/targetReference/text/@Image],
                        $target2 in //Flow/*[name/text/@Image=$target1//nextValueConnector/targetReference/text/@Image]
                            return $node[//Flow/actionCalls[name/text/@Image=$target2//nextValueConnector/targetReference/text/@Image]]
                    )
                    |
                    (: indirect reference, 3rd level :)
                    (for $node in //Flow/loops,
                        $target1 in //Flow/*[name/text/@Image=$node//nextValueConnector/targetReference/text/@Image],
                        $target2 in //Flow/*[name/text/@Image=$target1//nextValueConnector/targetReference/text/@Image],
                        $target3 in //Flow/*[name/text/@Image=$target2//nextValueConnector/targetReference/text/@Image]
                            return $node[//Flow/actionCalls[name/text/@Image=$target3//nextValueConnector/targetReference/text/@Image]]
                    )
                )/name
                ]]>
                </value>
            </property>
        </properties>
    </rule>
</ruleset>
  ```
</details>

It's crude, I don't like the fact I'm blanket treating all `actionCalls` as suspicious, there's `<actionType>apex</actionType>`... but it does detect what I want (and I included `recordLookups` too, [they can blow another limit](https://help.salesforce.com/s/articleView?id=sf.flow_considerations_limit_transaction.htm&type=5))

![image](https://user-images.githubusercontent.com/7957690/193431350-76509ca4-d1f6-4b2b-9384-7b896e4782ba.png)


Ideally there could be 1 more condition (parameter passed to the rule?) about the Status tag. For example to run only when `<status>Active</status>` (It can be [Active, Draft, Obsolete, InvalidDraft](https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/meta_visual_workflow.htm)). My current project has lots of flows that were rewritten but nobody has guts to actually hit the delete button.

Too much for my xpath-fu.

**Comment by rsoesemann** [comment]
Wow, @aidan-harding there happened a lot of discussion on this ticket since we first talked.

I had an idea that might simplify you goal to create an own language for flow to better detect issues than in the metadata XML from Salesforce. Maybe instead of having your own language and a defined AST, couldn't you just use an XSLT to transform into a better suited XML format. I found XSLT very powerful.

This way we could use standard XML rules with a flow pre-processor... Just a crazy midnight idea?!

**Comment by adangel** [comment]
> Ideally there could be 1 more condition (parameter passed to the rule?) about the Status tag. 

You can add another rule property for checking the status:

<details>

```xml
<ruleset
  xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  name="Salesforce Flow"
  xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 http://pmd.sourceforge.net/ruleset_2_0_0.xsd">
  <description>Salesforce Flow PMD ruleset, focused on performance issues</description>
  <exclude-pattern>.*/force-app/main/default/.*</exclude-pattern>
  <include-pattern>.*/force-app/main/default/flows/.*</include-pattern>

  <rule
      name="FlowWithSoqlDmlInLoop"
      language="xml"
      message="Avoid SOQL, DML in a loop in flows"
      class="net.sourceforge.pmd.lang.rule.XPathRule">
      <description/>
      <priority>1</priority>
      <properties>
          <property name="status" type="List[String]" value="Active|Draft" description="TODO"/>
          <property name="version" value="2.0"/>
          <property name="xpath">
              <value>
              <![CDATA[
              (
                  (: direct reference :)
                  //Flow/loops[nextValueConnector/targetReference/text/@Image=//Flow/(recordLookups|recordCreates|recordDeletes|recordUpdates)/name/text/@Image]
                  |
                  (: indirect reference :)
                  (for $node in //Flow/loops,
                      $target1 in //Flow/*[name/text/@Image=$node//nextValueConnector/targetReference/text/@Image]
                          return $node[//Flow/(recordLookups|recordCreates|recordDeletes|recordUpdates)[name/text/@Image=$target1//targetReference/text/@Image]]
                  )
                  |
                  (: indirect reference, 2nd level :)
                  (for $node in //Flow/loops,
                      $target1 in //Flow/*[name/text/@Image=$node//nextValueConnector/targetReference/text/@Image],
                      $target2 in //Flow/*[name/text/@Image=$target1//nextValueConnector/targetReference/text/@Image]
                          return $node[//Flow/(recordLookups|recordCreates|recordDeletes|recordUpdates)[name/text/@Image=$target2//targetReference/text/@Image]]
                  )
                  |
                  (: indirect reference, 3rd level :)
                  (for $node in //Flow/loops,
                      $target1 in //Flow/*[name/text/@Image=$node//nextValueConnector/targetReference/text/@Image],
                      $target2 in //Flow/*[name/text/@Image=$target1//nextValueConnector/targetReference/text/@Image],
                      $target3 in //Flow/*[name/text/@Image=$target2//nextValueConnector/targetReference/text/@Image]
                          return $node[//Flow/(recordLookups|recordCreates|recordDeletes|recordUpdates)[name/text/@Image=$target3//targetReference/text/@Image]]
                  )
              )/name[../../status/text/@Image = $status]
              ]]>
              </value>
          </property>
      </properties>
  </rule>
  <rule
      name="FlowWithApexInLoop"
      language="xml"
      message="Avoid Invocable Actions in a loop in flows"
      class="net.sourceforge.pmd.lang.rule.XPathRule">
      <description/>
      <priority>2</priority>
      <properties>
          <property name="status" type="List[String]" value="Active|Draft" description="TODO"/>
          <property name="version" value="2.0"/>
          <property name="xpath">
              <value>
              <![CDATA[
              (
                  (: direct reference :)
                  //Flow/loops[nextValueConnector/targetReference/text/@Image=//Flow/actionCalls/name/text/@Image]
                  |
                  (: indirect reference :)
                  (for $node in //Flow/loops,
                      $target1 in //Flow/*[name/text/@Image=$node//nextValueConnector/targetReference/text/@Image]
                          return $node[//Flow/actionCalls[name/text/@Image=$target1//targetReference/text/@Image]]
                  )
                  |
                  (: indirect reference, 2nd level :)
                  (for $node in //Flow/loops,
                      $target1 in //Flow/*[name/text/@Image=$node//nextValueConnector/targetReference/text/@Image],
                      $target2 in //Flow/*[name/text/@Image=$target1//nextValueConnector/targetReference/text/@Image]
                          return $node[//Flow/actionCalls[name/text/@Image=$target2//nextValueConnector/targetReference/text/@Image]]
                  )
                  |
                  (: indirect reference, 3rd level :)
                  (for $node in //Flow/loops,
                      $target1 in //Flow/*[name/text/@Image=$node//nextValueConnector/targetReference/text/@Image],
                      $target2 in //Flow/*[name/text/@Image=$target1//nextValueConnector/targetReference/text/@Image],
                      $target3 in //Flow/*[name/text/@Image=$target2//nextValueConnector/targetReference/text/@Image]
                          return $node[//Flow/actionCalls[name/text/@Image=$target3//nextValueConnector/targetReference/text/@Image]]
                  )
              )/name[../../status/text/@Image = $status]
              ]]>
              </value>
          </property>
      </properties>
  </rule>
</ruleset>
```

</details>

Crucial are these parts:

```xml
<property name="status" type="List[String]" value="Active|Draft" description="TODO"/>
```

This defines the property with a default value. In can have multiple values, delimited by `|`.


```
[../../status/text/@Image = $status]
```

This limits the results to flows which have one of the configured status. With the default value of the property "$status", only flows that are either "Active" or "Draft" are considered.


> It's crude, I don't like the fact I'm blanket treating all actionCalls as suspicious, there's `<actionType>apex</actionType>`...

You can limit it of course to detect only actionCalls of type apex - but you still don't know what the called method will do. So, this rule can't be 100% trustworthy. It can only point to you places in code, where you need to have closer look. We could define another property which whitelists/excludes specific actionNames - but we won't be able to provide generic default values for that.

**Comment by rsoesemann** [comment]
@aidan-harding should we close this one?

**Comment by mitchspano** [comment]
We’ve been able to show that XPath-based rules work well for finding problematic patterns in Salesforce Flow XML. Given that context, I suggest we close this proposal and take a different direction.

Introducing a domain-specific language (or a dedicated PMD language) to represent Flow would be difficult to maintain: by design it would always lag behind the canonical platform definition of Flow’s representation. As Salesforce evolves the Flow XML schema and semantics, we’d be constantly playing catch-up; a pain I am personally [very familiar](https://github.com/google/flow-lens/issues/39) with.

I think we should just begin by contributing a set of canonical XPath-based Flow rules to PMD so that:
- People have a clear, official starting point for writing their own Flow rules.
- The community can extend and adapt rules without maintaining a separate Flow “language” in PMD.
- Rule authors can target the actual Flow XML directly, avoiding an extra abstraction layer.

We also shouldn’t limit the scope of this effort to Flow. Many other metadata types could benefit from XPath-based rules: FlexiPage is an obvious next candidate, and we could even target `*-meta.xml` files that accompany every metadata type - ex: rules for LWC component property exposure, or API version rules for Apex. That would let us get very specific and support the Salesforce community across all facets of Salesforce development, not just Flow.

I think we should close this issue and focus on contributing well-scoped, canonical rules (starting with Flow) in a clearly designated Salesforce ruleset, with a path to include more metadata types over time.

**Comment by jsotuyod** [comment]
This is one of the reasons why we recently introduced dialects (see #5438, [release notes](https://docs.pmd-code.org/latest/pmd_release_notes_old.html#experimental-support-for-language-dialects)). Dialects provide exactly that, allows to build a specific set of rules (and even XPath functions) specific for a type of files, but sitting on top of the underlaying language infra and tools (ie: XML). WSDL, XSL and POM have already been reworked as XML dialects. 

The one big constraint we still currently have is automatic detection (see #4945), as the current implementation in PMD is limited to the file extension, so files ended in `.xml` will not match automatically (but can be worked around by using `--force-language <langdialect>`). For files using a specific extension (ie: `.flow`) this is not currently a problem.
